src/crafting/CraftingManager.php

```diff
@@ -214,6 +258,34 @@ public function registerShapelessRecipe(ShapelessRecipe $recipe) : void{
 		}
 	}
 
+	public function unregisterShapelessRecipe(ShapelessRecipe $recipe) : void{
+		$edited = false;
+		$hash = self::hashOutputs($recipe->getResults());
+
+		foreach($this->shapelessRecipes[$hash] ?? [] as $i => $r){
+			if($r === $recipe){
+				unset($this->shapelessRecipes[$hash][$i]);
+				if(count($this->shapelessRecipes[$hash]) === 0){
+					unset($this->shapelessRecipes[$hash]);
+					$edited = true;
+				}
+				break;
+			}
+		}
+
+		$index = array_search($recipe, $this->craftingRecipeIndex, true);
```

Comment:
If it wasn't for the `protected` fields, we could have just made the hash lookups store the index in here directly and avoided this :(

Not the fault of the PR, just an artifact of too much internals exposure.